tools/crm_mon.c

```diff
@@ -2807,18 +2853,363 @@ print_failed_actions(FILE *stream, pe_working_set_t *data_set)
 
     /* End section */
     switch (output_format) {
+        case mon_output_html:
+        case mon_output_cgi:
+            fprintf(stream, " </ul>\n");
+            break;
+
+        case mon_output_xml:
+            fprintf(stream, "    </failures>\n");
+            break;
+
+        default:
+            break;
+    }
+}
+
+/*!
+ * \internal
+ * \brief Reduce the stonith-history
+ *        for successful actions we keep the last of every action-type & target
+ *        for failed actions we record as well who had failed
+ *        for actions in progress we keep full track
+ *
+ * \param[in] history    List of stonith actions
+ *
+ */
+static stonith_history_t *
+reduce_stonith_history(stonith_history_t *history)
+{
+    stonith_history_t *new = NULL, *hp, *np, *tmp;
+
+    for (hp = history; hp; ) {
+        for (np = new; np; np = np->next) {
```

Comment:
More comments about how these loops work would be welcome. What entries are hp and np supposed to contain at any given time? Are they two pointers into a single list, or two lists, or what?